rfcs/0006-transaction.md

```diff
@@ -0,0 +1,300 @@
+
+This approach provides Serializable isolation level by detecting both Write-Write conflicts and Read-Write conflicts. It does not rely on MemTable size like RocksDB does, and provides more accurate conflict detection, without worrying about the MemTable size.
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn_opts = TransactionOptions::new().with_isolation_level(IsolationLevel::Serializable); // or IsolationLevel::Snapshot
+let txn = kv_store.transaction_with_opts(txn_opts).await.unwrap();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the data structure of the global `Oracle` and the transaction state:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
+    recent_committed_txns: Deque<Arc<TransactionState>>,
+}
+
+struct TransactionState {
+    started_seq: u64,
+    write_keys: HashSet<Bytes>,
+    read_keys: Vec<Range<Bytes>>,
+    committed_seq: Option<u64>
```

Comment:
> I think it's possible for a txn to complete, but for the write not to be durable yet 

i am a bit worried about having no D in ACID. 

also, i still feel a bit confused about the expected semantics of `await_durable`. It would be more natural for me to always await it to be durable, but users could choose to fire-and-forget at the caller site like spawning a task for this writing. 😲